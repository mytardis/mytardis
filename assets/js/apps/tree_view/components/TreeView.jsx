import React, { Fragment, useEffect, useState } from 'react';

import { decorators, Treebeard } from 'react-treebeard';

import PropTypes from 'prop-types';
import styles from './custom-theme';
import Header from './Header';
import Container from './Container';
import * as filters from './Filter';
import 'regenerator-runtime/runtime';
import { TreeDownloadButton, TreeSelectButton } from './Download';
import { DownloadArchive, FetchFilesInDir } from './Utils';


const TreeView = ({ datasetId, modified }) => {
  const [cursor, setCursor] = useState(false);
  const [data, setData] = useState([]);
  const [selectedCount, setSelectedCount] = useState(0);
  const [isAllSelected, setIsAllSelected] = useState(false);
  const onSelect = (node) => {
    node.toggled = !node.toggled;
    if (node.selected) {
      // deselect all child nodes
      // if this is a folder and has child nodes
      if (node.children && node.children.length) {
        node.selected = false;
        node.children.forEach((childNode) => {
          childNode.selected = false;
        });
      } else {
        node.selected = false;
        setSelectedCount(selectedCount - 1);
      }
    } else {
      node.selected = true;
      // if this is a folder with no child
      if (node.children && !node.children.length) {
        // add this node to selecteNode list
        node.selected = true;
      }
      // if this is a folder with child
      if (node.children && node.children.length) {
        node.children.forEach((childNode) => {
          childNode.selected = true;
        });
      } else {
        node.selected = true;
      }
    }
    // set selected count
    // set count
    let count = 0;
    data.forEach((item) => {
      const selectedItemCount = filters.countSelection(item, 0);
      count += selectedItemCount;
    });
    setSelectedCount(count);
  };
  const fetchBaseDirs = () => {
    fetch(`/api/v1/dataset/${datasetId}/root-dir-nodes/`, {
      method: 'get',
      headers: {
        'Accept': 'application/json', // eslint-disable-line quote-props
        'Content-Type': 'application/json',
      },
    }).then(responseJson => (responseJson.json()))
      .then((response) => { setData(response); });
  };
  const fetchChildDirs = (node, dirPath) => {
    const encodedDir = encodeURIComponent(dirPath);
    fetch(`/api/v1/dataset/${datasetId}/child-dir-nodes/?dir_path=${encodedDir}`, {
      method: 'get',
      headers: {
        'Accept': 'application/json', // eslint-disable-line quote-props
        'Content-Type': 'application/json',
      },
    }).then(response => (response.json()))
      .then((childNodes) => {
        node.children = childNodes;
        node.toggled = true;
        if (node.selected) {
          let childCount = 0;
          node.children.forEach((childNode) => {
            childNode.toggled = true;
            onSelect(childNode);
            childCount += 1;
          });
          setSelectedCount(selectedCount + childCount);
        }
        setData(Object.assign([], data));
      });
    //
  };
  useEffect(() => {
    fetchBaseDirs('');
  }, [datasetId, modified]);
  const onToggle = (node, toggled) => {
    // fetch children:
    // console.log(`on toggle ${node.name}`);
    if (toggled && node.children && node.children.length === 0) {
      fetchChildDirs(node, node.path);
      return;
    }
    if (cursor) {
      cursor.active = false;
    }
    node.active = true;
    if (node.children) {
      node.toggled = toggled;
    }
    setCursor(node);
    setData(Object.assign([], data));
  };
  const onFilterMouseUp = ({ target: { value } }) => {
    const filter = value.trim();
    if (!filter) {
      // set initial tree state:
      fetchBaseDirs('');
      // set count to 0
      setSelectedCount(0);
    }
    const filteredData = [];
    data.forEach((item) => {
      let filtered = filters.filterTree(item, filter);
      filtered = filters.expandFilteredNodes(filtered, filter);
      filteredData.push(filtered);
    });

    setData(filteredData);
  };

  const downloadSelected = (event) => {
    event.preventDefault();
    const formData = new FormData(event.target);
    let selectedData = [];
    // get selected files and folders
    data.forEach((item) => {
      const selected = filters.findSelected(item, []);
      selectedData = [...selectedData, ...selected];
    });
    // if folder is selected, get array of promises
    // This will resolve to a list of ids for each folder
    let selectedFileIds = [];
    function getFilesFromDir() {
      const promises = [];
      for (let i = 0; i < selectedData.length; i += 1) {
        if (!Array.isArray(selectedData[i])) {
          if (!selectedData[i].id) {
            const encodedDir = encodeURIComponent(selectedData[i].path);
            promises.push(FetchFilesInDir(datasetId, encodedDir));
          } else {
            selectedFileIds = [...selectedFileIds, selectedData[i].id.toString()];
          }
        }
      }
      return Promise.all(promises);
    }
    getFilesFromDir().then((responses) => {
      responses.forEach((resp) => {
        selectedFileIds = [...selectedFileIds, ...resp];
      });
      return selectedFileIds;
    }).then((selectedIds) => {
      selectedIds.forEach((id) => {
        formData.append('datafile', id.toString());
      });
      DownloadArchive(formData);
    });
  };
  const toggleSelection = (event) => {
    event.preventDefault();
    // if count < 1 select all else select None
    const selectedData = [];
    if (!isAllSelected) {
      data.forEach((item) => {
        const selectedNodes = filters.toggleSelection(item, true);
        selectedData.push(selectedNodes);
      });
    } else {
      data.forEach((item) => {
        const selectedNodes = filters.toggleSelection(item, false);
        selectedData.push(selectedNodes);
      });
    }
    // set data
    setData(selectedData);
    // set all selected to true
    setIsAllSelected(!isAllSelected);
    // set count
    let count = 0;
    selectedData.forEach((item) => {
      const selectedItemCount = filters.countSelection(item, 0);
      count += selectedItemCount;
    });
    setSelectedCount(count);
  };
  return (
    <Fragment>
      <div>
        <div style={{ float: 'left' }}>
          <TreeDownloadButton
            count={selectedCount}
            onClick={downloadSelected}
          />
        </div>
        <div>
          <TreeSelectButton
            count={selectedCount}
            onClick={toggleSelection}
            buttonText={isAllSelected ? 'Select None' : 'Select All'}
          />
        </div>
      </div>
      <div style={styles}>
        <div className="input-group">
          <span className="input-group-text">
            <i className="fa fa-search" />
          </span>
          <input
            name="search-input"
            className="form-control"
            onKeyUp={onFilterMouseUp}
            placeholder="Search the tree..."
            type="text"
          />
        </div>
      </div>
      <div className="p-1" style={styles}>
        <Treebeard
          data={data}
          style={styles}
          onToggle={onToggle}
          onSelect={onSelect}
          decorators={{ ...decorators, Header, Container }}
          animation={false}
        />
      </div>
    </Fragment>
  );
};

TreeView.propTypes = {
  datasetId: PropTypes.string.isRequired,
  modified: PropTypes.string,
};

TreeView.defaultProps = {
  modified: '',
};

export default TreeView;
