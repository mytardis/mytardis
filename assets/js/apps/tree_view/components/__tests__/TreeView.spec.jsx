import React from 'react';
import ReactDOM from 'react-dom';
import { act } from '@testing-library/react';
import { configure, mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import TreeView from '../TreeView';


global.fetch = require('jest-fetch-mock');

configure({ adapter: new Adapter() });

const fakeTreeData = [{
  name: 'parent_1',
  path: 'parent_1',
  children: [],
},
{
  name: 'parent_2',
  path: 'parent_2',
  children: [],
},
{
  name: 'Parent.txt',
  id: 11985776,
  verified: true,
},
{
  name: 'STORM-6.jpg',
  id: 11985840,
  verified: false,
}];
const fakeChildData = [
  {
    name: 'child_1',
    path: 'parent_1/child_1',
    children: [],
  },
  {
    name: 'child_2',
    path: 'parent_1/child_2',
    children: [],
  },
  {
    name: 'Child_1.txt',
    id: 11985763,
    verified: true,
  },
  {
    name: 'Child_2.txt',
    id: 11985764,
    verified: true,
  }];

let container = null;
let component = null;
beforeEach(async () => {
  jest.spyOn(global, 'fetch').mockImplementation(() => Promise.resolve({
    json: () => Promise.resolve(fakeTreeData),
  }));
  container = document.createElement('div');
  document.body.appendChild(container);
  await act(async () => {
    component = mount(<TreeView datasetId="1234" modified="" />, { attachTo: container });
  });
  component.update();
});
afterEach(() => {
  // cleanup on exiting
  container.remove();
  container = null;
  component.unmount();
  component.detach();
});

describe('renders initial tree view on page load', () => {
  it('should render folder tree correctly', async () => {
    await act(async () => {
      ReactDOM.render(<TreeView datasetId="1234" modified="" />, container);
    });
    expect(container).toMatchSnapshot();
  });
  it('should render tree view with four child nodes', async () => {
    await act(async () => {
      ReactDOM.render(<TreeView datasetId="1234" modified="" />, container);
    });
    expect(container.querySelector('ul').children.length).toEqual(4);
    expect(container.querySelector('ul').children[0].textContent).toEqual('parent_1');
    expect(container.querySelector('ul').children[1].textContent).toEqual('parent_2');
    expect(container.querySelector('ul').children[2].textContent).toEqual('Parent.txt');
    expect(container.querySelector('ul').children[3].textContent).toEqual('STORM-6.jpg(unverified)');
  });
});

describe('test filter, select and toggle node', () => {
  afterAll(() => {
    component.unmount();
    component.detach();
    component.update();
  });
  it('should render child nodes when clicked on parent', async () => {
    jest.spyOn(global, 'fetch').mockImplementation(() => Promise.resolve({
      json: () => Promise.resolve(fakeChildData),
    }));
    // select folder before toggle
    await act(async () => {
      const checkBox = component.find({ type: 'checkbox' }).first();
      checkBox.simulate('click');
      setImmediate(() => {
        component.update();
        expect(component.find('Header').get(0).props.node.selected).toBeTruthy();
      });
    });
    // toggle folder open
    await act(async () => {
      component.find('NodeHeader').first().simulate('click');
    });
    component.update();
    expect(component.find('NodeHeader')).toHaveLength(8);
    expect(component.find('Header').get(0).props.iconClass).toEqual('folder-open');
    expect(component.find('Header').get(0).props.node.children.length).toEqual(4);
    expect(component.find('Header').get(0).props.node.children[0].name).toEqual('child_1');
    expect(component.find('TreeDownloadButton').get(0).props.count).toEqual(5);
    // toggle folder close
    await act(async () => {
      component.find('NodeHeader').first().simulate('click');
    });
    component.update();
    expect(component.find('NodeHeader')).toHaveLength(4);
    expect(component.find('Header').get(0).props.iconClass).toEqual('folder');
  });
  it('should filter tree node on search text change', async () => {
    await act(async () => {
      const searchField = component.find('input.form-control');
      searchField.simulate('keyUp', { target: { name: 'search-input', value: '2' } });
      // searchField.simulate('keyUp', { keyCode: 50 });
      setImmediate(() => {
        component.update();
        expect(component.find('NodeHeader')).toHaveLength(6);
      });
    });
  });
  it('should reload the tree if search text is empty', async () => {
    jest.spyOn(global, 'fetch').mockImplementation(() => Promise.resolve({
      json: () => Promise.resolve(fakeTreeData),
    }));
    await act(async () => {
      const searchField = component.find('input.form-control');
      searchField.simulate('keyUp', { target: { name: 'search-input', value: '' } });
      setImmediate(() => {
        component.update();
        expect(component.find('NodeHeader')).toHaveLength(4);
        // should set count to 0
        expect(component.find('TreeDownloadButton').get(0).props.count).toEqual(0);
      });
    });
  });
  it('should deselect a node', async () => {
    await act(async () => {
      const checkBox = component.find({ type: 'checkbox' }).first();
      checkBox.simulate('click');
    });
    component.update();
    expect(component.find('Header').get(0).props.node.selected).toBeFalsy();
  });
  it('should select all child nodes, if parent node is selected', async () => {
    expect(component.find('Header').get(0).props.node.selected).toBeFalsy();
    // select parent Node
    await act(async () => {
      const checkBox = component.find({ type: 'checkbox' }).first();
      checkBox.simulate('click');
    });
    component.update();
    expect(component.find('Header').get(0).props.node.selected).toBeTruthy();
    jest.spyOn(global, 'fetch').mockImplementation(() => Promise.resolve({
      json: () => Promise.resolve(fakeChildData),
    }));
    await act(async () => {
      component.find('Container').first().simulate('click');
    });
    component.update();
    expect(component.find('NodeHeader')).toHaveLength(8);
    expect(component.find('Header').get(0).props.iconClass).toEqual('folder-open');
    expect(component.find('Header').get(1).props.node.selected).toBeTruthy();
    expect(component.find('Header').get(2).props.node.selected).toBeTruthy();
    component.update();
    expect(component.find('TreeDownloadButton').get(0).props.count).toEqual(5);
  });
  it('should deselect folder nodes', async () => {
    expect(component.find('Header').get(1).props.node.selected).toBeTruthy();
    await act(async () => {
      const checkBox = component.find({ type: 'checkbox' }).at(1);
      checkBox.simulate('click');
    });
    component.update();
    expect(component.find('NodeHeader')).toHaveLength(8);
    expect(component.find('Header').get(1).props.node.selected).toBeFalsy();
  });
  it('should select  a folder nodes', async () => {
    expect(component.find('Header').get(1).props.node.selected).toBeFalsy();
    await act(async () => {
      const checkBox = component.find({ type: 'checkbox' }).at(1);
      checkBox.simulate('click');
    });
    component.update();
    expect(component.find('NodeHeader')).toHaveLength(8);
    expect(component.find('Header').get(1).props.node.selected).toBeTruthy();
  });
  it('should deselect all child nodes, if parent node is deselected', async () => {
    await act(async () => {
      const checkBox = component.find({ type: 'checkbox' }).first();
      checkBox.simulate('click');
    });
    component.update();
    expect(component.find('NodeHeader')).toHaveLength(8);
    expect(component.find('Header').get(1).props.node.selected).toBeFalsy();
    expect(component.find('Header').get(2).props.node.selected).toBeFalsy();
  });
  it('should select child node with no children', async () => {
    await act(async () => {
      const checkBox = component.find({ type: 'checkbox' }).at(6);
      checkBox.simulate('click');
    });
    component.update();
    expect(component.find('NodeHeader')).toHaveLength(8);
    expect(component.find('Header').get(6).props.node.selected).toBeTruthy();
  });
  it('should deselect child node with no children', async () => {
    expect(component.find('Header').get(6).props.node.selected).toBeTruthy();
    await act(async () => {
      const checkBox = component.find({ type: 'checkbox' }).at(6);
      checkBox.simulate('click');
    });
    component.update();
    expect(component.find('NodeHeader')).toHaveLength(8);
    expect(component.find('Header').get(6).props.node.selected).toBeFalsy();
  });
  it('should select all nodes when select All button is clicked', async () => {
    // find select all button
    await act(async () => {
      const selectAllButton = component.find('button').at(1);
      selectAllButton.simulate('click');
    });
    component.update();
  });
  it('should deselect all nodes when select None button is clicked', async () => {
    // find select all button
    await act(async () => {
      const selectAllButton = component.find('button').at(1);
      selectAllButton.simulate('click');
    });
    component.update();
    await act(async () => {
      const selectAllButton = component.find('button').at(1);
      selectAllButton.simulate('click');
    });
    component.update();
  });
});

describe('test download selected files', () => {
  it('should download selected files', async () => {
    // select 2 files at root of the tree
    await act(async () => {
      const checkBox = component.find({ type: 'checkbox' }).at(3);
      checkBox.simulate('click');
    });
    component.update();
    await act(async () => {
      const checkBox = component.find({ type: 'checkbox' }).at(6);
      checkBox.simulate('click');
    });
    component.update();
    // click on Download
    fetch.mockResponseOnce('[\'a\', \'b\', \'c\', \'d\']', {
      status: 200,
      headers: {
        'content-type': 'application/x-tar',
        'Content-Disposition': 'attachment; filename=manish-selection.tar',
        'Content-Length': '71680',
      },
    });
    jest.mock('file-saver', () => ({ saveAs: jest.fn() }));
    global.Blob = function f(content, options) { return ({ content, options }); };
    global.URL.createObjectURL = jest.fn();
    const downloadButton = component.find('form').simulate('submit');
    await act(async () => {
      downloadButton.simulate('click');
    });
    component.update();
    expect(fetch.mock.calls.length).toEqual(20);
    expect(fetch.mock.calls[19][1].body.get('comptype')).toEqual('tar');
    expect(fetch.mock.calls[19][1].body.get('organization')).toEqual('deep-storage');
    expect(fetch.mock.calls[19][1].body.getAll('datafile')).toEqual(['11985763', '11985776']);
  });
  it('should download all files within a selected folder', async () => {
    // select Parent_1/child_1 folder
    await act(async () => {
      const checkBox = component.find({ type: 'checkbox' }).at(1);
      checkBox.simulate('click');
    });
    component.update();
    // eslint-disable-next-line global-require
    const utils = require('../Utils');
    utils.FetchFilesInDir = jest.fn(() => [11985777]);
    fetch.mockResponseOnce('[\'a\', \'b\', \'c\', \'d\']', {
      status: 200,
      headers: {
        'content-type': 'application/x-tar',
        'Content-Disposition': 'attachment; filename=manish-selection.tar',
        'Content-Length': '71680',
      },
    });
    jest.mock('file-saver', () => ({ saveAs: jest.fn() }));
    global.Blob = function f(content, options) { return ({ content, options }); };
    global.URL.createObjectURL = jest.fn();
    const downloadButton = component.find('form').simulate('submit');
    await act(async () => {
      downloadButton.simulate('click');
    });
    component.update();
    expect(fetch.mock.calls.length).toEqual(22);
    // expect download datafile endpoint is called
    expect(fetch.mock.calls["21"][0]).toEqual('/download/datafiles/');
    // expect methos to be POST
    expect(fetch.mock.calls["21"][1].method).toEqual('POST');
    // expect form data to include 3 datafile
    expect(fetch.mock.calls['21']['1'].body.getAll('datafile')).toEqual(['11985763', '11985776', '11985777']);
  });
  it('should call api to get files in subdir', () => {
    jest.clearAllMocks().resetModules();
    // eslint-disable-next-line global-require
    const utils = require('../Utils');
    jest.spyOn(global, 'fetch').mockImplementation(() => Promise.resolve({
      json: () => Promise.resolve(12345),
    }));
    // eslint-disable-next-line no-unused-vars
    const response = utils.FetchFilesInDir('1234', 'Parent_1/');
    expect(fetch.mock.calls.length).toEqual(1);
    expect(fetch).toHaveBeenCalledWith('/api/v1/dataset/1234/child-dir-files/?dir_path=Parent_1/');
  });
});
