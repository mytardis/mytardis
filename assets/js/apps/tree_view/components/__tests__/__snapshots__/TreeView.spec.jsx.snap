// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders initial tree view on page load should render folder tree correctly 1`] = `
<div>
  <div>
    <div
      style="float: left;"
    >
      <form>
        <input
          type="hidden"
        />
        <input
          name="comptype"
          type="hidden"
          value="tar"
        />
        <input
          name="organization"
          type="hidden"
          value="deep-storage"
        />
        <button
          class="btn btn-outline-secondary"
          disabled=""
          style="margin-bottom: 12px; font-weight: bold; margin-right: 2px;"
          title="Start selecting files to download"
          type="submit"
          value="submit"
        >
          <i
            class="fa fa-download fa-sm mr-1 "
          />
          Select Files...
          
        </button>
      </form>
    </div>
    <div>
      <button
        class="btn btn-outline-secondary"
        style="margin-bottom: 12px; font-weight: bold;"
        title="Select All"
        type="submit"
        value="submit"
      >
        Select All
      </button>
    </div>
  </div>
  <div>
    <div
      class="input-group"
    >
      <span
        class="input-group-text"
      >
        <i
          class="fa fa-search"
        />
      </span>
      <input
        class="form-control"
        name="search-input"
        placeholder="Search the tree..."
        type="text"
      />
    </div>
  </div>
  <div
    class="p-1"
  >
    <ul
      class="css-1tl97mg"
    >
      <li
        class="css-ya4lz1"
      >
        <div>
          <div
            style="display: inline-block; vertical-align: top; color: rgb(0, 0, 0); padding-bottom: 2px;"
          >
            <div
              style="line-height: 20px; vertical-align: middle; font-size: 16px; cursor: pointer;"
            >
              <input
                class="datafile_checkbox"
                readonly=""
                style="margin-right: 5px;"
                type="checkbox"
              />
              <i
                class="fa fa-folder"
                style="margin-right: 5px; opacity: 0.6;"
              />
              parent_1
              
            </div>
          </div>
        </div>
        <div />
      </li>
      <li
        class="css-ya4lz1"
      >
        <div>
          <div
            style="display: inline-block; vertical-align: top; color: rgb(0, 0, 0); padding-bottom: 2px;"
          >
            <div
              style="line-height: 20px; vertical-align: middle; font-size: 16px; cursor: pointer;"
            >
              <input
                class="datafile_checkbox"
                readonly=""
                style="margin-right: 5px;"
                type="checkbox"
              />
              <i
                class="fa fa-folder"
                style="margin-right: 5px; opacity: 0.6;"
              />
              parent_2
              
            </div>
          </div>
        </div>
        <div />
      </li>
      <li
        class="css-ya4lz1"
      >
        <div>
          <div
            style="display: inline-block; vertical-align: top; color: rgb(0, 0, 0); padding-bottom: 2px;"
          >
            <div
              style="line-height: 20px; vertical-align: middle; font-size: 16px; cursor: pointer;"
            >
              <input
                class="datafile_checkbox"
                readonly=""
                style="margin-right: 5px;"
                type="checkbox"
              />
              <i
                class="fa fa-file-text"
                style="margin-right: 5px; opacity: 0.6;"
              />
              Parent.txt
              <a
                class="btn-sm"
                href="/api/v1/dataset_file/11985776/download/"
                style="color: black;"
                title="Download"
              >
                <i
                  class="fa fa-download fa-sm"
                />
              </a>
            </div>
          </div>
        </div>
        <div />
      </li>
      <li
        class="css-ya4lz1"
      >
        <div>
          <div
            style="display: inline-block; vertical-align: top; color: rgb(0, 0, 0); padding-bottom: 2px;"
          >
            <div
              style="line-height: 20px; vertical-align: middle; font-size: 16px; cursor: pointer;"
            >
              <input
                class="datafile_checkbox"
                disabled=""
                readonly=""
                style="margin-right: 5px;"
                type="checkbox"
              />
              <i
                class="fa fa-file-text"
                style="margin-right: 5px; opacity: 0.6;"
              />
              <span
                style="color: red;"
              >
                STORM-6.jpg
                (unverified)
              </span>
              <a
                class="btn-sm"
                href="/api/v1/dataset_file/11985840/download/"
                style="pointer-events: none; cursor: default; text-decoration: none; color: grey;"
                title="Download"
              >
                <i
                  class="fa fa-download fa-sm"
                />
              </a>
            </div>
          </div>
        </div>
        <div />
      </li>
    </ul>
  </div>
</div>
`;
